mlir/lib/Dialect/MIOpen/utility/transformMapUtils.cpp

```diff
@@ -0,0 +1,413 @@
+// General utilities.
+//===----------------------------------------------------------------------===//
+std::tuple<Value, ArrayAttr>
+mlir::miopen::untransform(OpBuilder &b, Value transformed, ArrayAttr existing) {
+  SmallVector<Attribute> transformList;
+  if (existing)
+    transformList.append(existing.begin(), existing.end());
+  Value ret = transformed;
+  while (auto transform = dyn_cast_or_null<TransformOp>(ret.getDefiningOp())) {
+    llvm::copy(transform.transforms(), std::back_inserter(transformList));
+    ret = transform.input();
+  }
+  return {ret, b.getArrayAttr(transformList)};
+}
+
+std::tuple<Value, ArrayAttr>
+mlir::miopen::untransform(OpBuilder &b, Value transformed,
+                          ArrayRef<Attribute> existing) {
+  return untransform(b, transformed, b.getArrayAttr(existing));
+}
+
+TransformOp mlir::miopen::reshapeBuffer(OpBuilder &b, Location loc,
+                                        Value buffer, ArrayRef<StringRef> names,
+                                        ArrayRef<int64_t> shape) {
+  MemRefType bufferType = buffer.getType().cast<MemRefType>();
+  ArrayRef<int64_t> outShape = bufferType.getShape();
+  assert(outShape.size() == 1 && "Buffer being reshaped must start linear");
+
+  SmallVector<int64_t> strides;
+  strides.reserve(shape.size());
+  int64_t stride = 1;
+  for (int64_t v : llvm::reverse(shape)) {
+    strides.push_back(stride);
+    stride *= v;
+  }
+  std::reverse(strides.begin(), strides.end());
+  assert(stride == outShape[0] && "Strides must multiply to buffer length");
+
+  TopDownTMBuilder transform(b, names, shape, loc);
+  transform.embed("raw", 0, outShape[0], names, strides);
+
+  TransformMapAttr transformAttr = transform.get();
+  auto ret = b.create<TransformOp>(loc, buffer, transformAttr,
+                                   bufferType.getMemorySpaceAsInt());
+  return ret;
+}
+
+//===----------------------------------------------------------------------===//
+// Vectorization inference.
+//===----------------------------------------------------------------------===//
+// A map of dimensions to the maximal length in those dimensions that can
+// be vectorized. If a dimension's maximal length is `None`, that dimension
+// is assumed to be held constant.
+using VectorizationData = llvm::IndexedMap<Optional<int64_t>>;
+
+static VectorizationData
+propagateVectorizationInfo(TransformMapAttr map,
+                           const VectorizationData &input) {
+  VectorizationData result;
+  result.grow(map.getMap().getValue().getNumResults() - 1);
+  for (TransformAttr transform : map.getOps()) {
+    ArrayRef<uint32_t> upperDims = transform.getUpperDims();
+    ArrayRef<uint32_t> lowerDims = transform.getLowerDims();
+    ArrayRef<int64_t> params = transform.getParams();
+
+    switch (transform.getType()) {
+    case TransformType::PassThrough:
+    case TransformType::AddDim:
+    case TransformType::Slice:
```

Comment:
For slice, if the upper dimension is 16 x 16 with vectorization size of 4 and lowering dimension is 15 x 15, can lower dimension continue to vectorize at length of 4? I'm assuming not because it crossed the boundary between.